### DISCUSSION TITLE
CI/CD Workflows

### DISCUSSION THREAD

**Discussion by zyx-billy** [MLIR-China/mlir-playground#32] [Ideas]
Proposal: Ordered from most common to least common.

- Build & upload docker image (#20)
  - Setup on github.
  - Runs on `main` and all PRs for `main`, and only if `./cpp/docker/` changed.
  - Builds the `clang-wasm` image, and pushes built image to dockerhub and tag with sha.
- Test website build
  - Setup on github.
  - Runs on all PRs.
  - Depends on previous action (build & upload docker image) succeeding (or skipped).
  - Runs the build script and make sure npm build succeeds.
  - Runs any tests we have.
- Deploy to staging
  - Setup on cloudflare.
  - Runs on `main`.
  - Steps:
    - Upload large files to b2 staging folder.
    - Build static website and serve from `./out/`.
- Deploy to production
  - Setup on cloudflare.
  - Runs on `deploy`.
  - Steps:
    - Upload large files to b2 production folder.
    - Build static website and serve from `./out/`.

**Comment by litanlitudan** [comment]
Do we have an agreement on which CDN vendor to build on? Other than that, everything else looks great to me. @zyx-billy

**Reply by zyx-billy** [reply]
Oh yes, we still need to settle on a cdn for next meeting. This is mainly for serving the large build files. If we can find a chinese alternative to cloudflare + b2 we can consider moving over.
